### PR TITLE
Fix aggregate_downsample() performance degradation

### LINKED ISSUE #13058: timeseries aggregate_downsample() appears to be significantly slower in Astropy 5
### Description

TimeSeries `aggregate_downsample()` appears to be significantly slower in Astropy 5 (comparing it with Astropy 4).

For the sample TimeSeries below (20000 data points), doing a `aggregate_downsample(ts, time_bin_size=10*u.minute)` , the elapsed time is roughtly:

- Astropy 4.3.1, Windows 11: < 1 sec
- Astropy 5, Windows 11: ~ 10 sec
- Astropy 5, Windows 11 with WSL2: ~ 7 sec

The astropy in question are installed with `conda`  (`pip` for WSL2).

### Steps to Reproduce
The sample time series and the timing script:

```python
# Test timeseries aggregate_downsample's running time

from astropy.time import Time
from astropy.timeseries import TimeSeries
from astropy import units as u
import numpy as np
from timeit import default_timer as timer

# the scale of a typical TESS 2-minute cadence lightcurve
num_points = 20000
time = Time(2457000 + np.arange(0, num_points) / 24 / 60 * 2, format="jd")
flux = np.ones(num_points)

ts = TimeSeries(time=time, data=dict(flux=flux))

start = timer()
ts_b = astropy.timeseries.aggregate_downsample(ts, time_bin_size=10*u.minute)
end = timer()

print("ts.bin(10 min) elapsed time:", (end - start))
len(ts), len(ts_b)
```

#### Windows 11 + Astropy 4.3.1
```
Windows-10-10.0.22000-SP0
Python  3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:21:54) [MSC v.1929 64 bit (AMD64)]
astropy  4.3.1
Numpy  1.22.3
pyerfa  2.0.0.1
Scipy  1.8.0
```

#### Windows 11 + Astropy 5
```
Windows-10-10.0.22000-SP0
Python  3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:21:54) [MSC v.1929 64 bit (AMD64)]
astropy  5.0.4
Numpy  1.22.3
pyerfa  2.0.0.1
Scipy  1.8.0
```

#### Windows 11 WSL2 + Astropy 5
```
Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.29
Python  3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0]
astropy  5.0.2
Numpy  1.22.3
pyerfa  2.0.0.1
Scipy  1.8.0
```

### PR DESCRIPTION
**Pull request by orionlee**
### Description

This pull request is to address performance degradatation of timeseries `aggregate_downsample()` in astropy5. 

For the test stated in the issue #13058, the elapsed time is reduced from ~10sec to ~1.5sec with this PR.

Fixes  #13058

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by orionlee** [comment]
Concern so far. If someone is better equipped to push forward, feel free to takeover. 

1. The fix is done by passing plain array to `np` functions with `.value` instead of `Time` objects, e.g., 

``` python
    indices = np.searchsorted(bin_end.value, ts_sorted.time[keep].value)
```

- 2 `Time.value` instances are not always comparable, if they are in different format, e.g,
  - time1 is of type `isot` , e.g., '2016-03-22T12:30:31', and time2 is `jd` (number)
  - time1 is `jd` , and time2 is `mjd`
  - I'm unsure what'd be a good way to resolve this. Do we convert the `Time` objects to a single format?

---

2. The PR failed 1 assertion in `test_downsample()`. I cannot figure out what the problem is.

https://github.com/astropy/astropy/blob/f54df76430eac0f67efb1e61be5a31f4c3029b9d/astropy/timeseries/tests/test_downsample.py#L127

```python
>       assert_equal(down_1["a"].data.data, np.array([1, 2, 3, 4, 5]))
E       AssertionError:
E       Arrays are not equal
E
E       Mismatched elements: 4 / 5 (80%)
E       Max absolute difference: 5
E       Max relative difference: 1.
E        x: array([1, 3, 4, 5, 0])
E        y: array([1, 2, 3, 4, 5])

astropy\timeseries\tests\test_downsample.py:127: AssertionError
```

**Comment by dhomeier** [comment]
Thanks for working on this!

> * I'm unsure what'd be a good way to resolve this. Do we convert the `Time` objects to a single format?

Yes, working on a common scale is mandatory, but range and precision has also to be taken into consideration. One might just use `time.jd`, `time.mjd` or `time.unix` to work with, but timeseries may well contain data at millisecond resolution, which would already hit `float64` resolution limits in these cases (`Time.now().jd` ~ 2.1e14 ms; for `unix` it is a bit better, but still not ideal).
I think the safest way will be to use all times relative to the first point in the series, just as done in the previous implementation:
https://github.com/aaryapatil/astropy/blob/cbf4d021b0e4b5ec9aa804cf4ce1145c19e0cf84/astropy/timeseries/downsample.py#L77

This will require some more bookkeeping, but for just the `ts_sorted` and `time_bin_start`/`_end` values hopefully does not become too complicated.

**Comment by orionlee** [comment]
@dhomeier Would it make sense to encapsulate all such logic (creating relative time and sort them) in a new `Time.searchsorted(v)` function (assuming that making `np.searchsorted()` run fast with `Time` is non-trivial)?

**Comment by dhomeier** [comment]
> @dhomeier Would it make sense to encapsulate all such logic (creating relative time and sort them) in a new `Time.searchsorted(v)` function (assuming that making `np.searchsorted()` run fast with `Time` is non-trivial)?

Given that there is a speed-up of O(1.e3) involved, I think a general fast search sorted for Time would certainly be worth it.
I don't know exactly what it would involve to make this available through `np.searchsorted`, but it looks like it could build on #12703. The logic for changing to relative and identical timescales would still need to be written, but that should be about the same effort as putting it into a `Time.searchsorted()` – note also @mhvk's comments on precision in that PR.
So I think once @dstansby's PR is merged this could go into a follow-up, implementing `searchsorted` in `CUSTOM_FUNCTIONS` analogously to `linspace`.

**Comment by orionlee** [comment]
This PR is ready, pending review and changelog.

**Comment by orionlee** [comment]
The `delete_indices = np.where(np.logical_and(ts_sorted.time > bin_end[ind], ts_sorted.time < bin_start[ind+1]))` does pose some problem, especially for large data set.
- if I replace `Time` object comparison with plain ndarray by using `.value`, it cuts down some time
- if we remove the check altogether (over `n_bins` times per `downsample` call), the time is further reduced to be quite close to the time in astropy 4.x 

Here is the breakdown using the test script in #13058
|                                                | 20k samples | 200k samples |   |
| ---------------------------------------------- | ----------- | ------------ | -- | 
| astropy 4.3.1                                  | ~0.42sec    | ~4.2sec      |  |
| astropy 5.0.4                                  | ~10sec      | ~252sec      | |
| searchsorted() fix (this PR)                   | ~1.3sec     | 147sec       | <-- **still very slow!!** | 
| `delete_indices` : use ndarray rather than `Time` for comparison | ~1.0sec     | ~14sec       | |
| remove `delete_indices` logic                   | ~0.55sec    | ~5sec        | |

---

I have done [a separate test](https://gist.github.com/orionlee/6b3e384b0b272cd1384a384ead7bde7b) for the running time of `Time` comparison logic. With 200k samples and 40k bins,
- `Time` object comparison: ~71sec
- `ndarray` comparison: ~3.7seec

--

For this PR
- I could replace `Time` comparison with ndarray, though it seems to be something that could better be done with `Time` itself.
- `delete_indices` check logic would remain to be a source of noticeable slow down.

**Comment by dhomeier** [comment]
Given the large speedup (I only saw something like a factor 2, but only experimented with referencing `Time.unix` directly in the loop), it should be worth defining `bin_start_sec = (bin_start - ts_sorted.time[0])to_value(u.s)` etc. outside of the loop, and then use the optimised loop as `np.where(bin_start_sec[1:] > bin_end_sec[:-1])[0]` (`+ epsilon`).
And one could think about setting something like `1e-3 * (ts_sorted.time[1:]-ts_sorted.time[:-1]).min()` as the epsilon to avoid most of those cases entirely.

**Comment by orionlee** [comment]
The PR is ready for review. See the table below for the running time improvement.

I'd wait for feedback before proceeding further.

It has 3 parts:
1. `searchsorted()`: replace `Time` objects with `ndarray(np.longdouble)` (of relative time)
2. reduce looping for handling noncontiguous bins (in deriving `delete_indices` and `keep`)
3. replace Time comparison with `ndarray(np.longdouble)` comparison with relative time (in deriving `delete_indices` and `keep`)

- (1) and (3) has the caveat that there is some loss on precision ( < 1ns on 64bit OS.)
- if we agree with the approach, there are rooms for further improvements, as there are more instances of `Time` comparison in the codes.


|                                                                            | 20k samples | 200k samples | Commit                                   | Comment                                                                                   |
| -------------------------------------------------------------------------- | ----------- | ------------ | ---------------------------------------- | ----------------------------------------------------------------------------------------- |
| astropy 4.3.1                                                              | ~0.42sec    | ~4.2sec      | --                                      |                                                                                           |
| astropy 5.0.4                                                              | ~10sec      | ~252sec      | --                                      |                                                                                           |
| `searchsorted()` fix (this PR initially)                                 | ~1.3sec     | 147sec       | a753502f60fab3e3631c8b0e69efc4a0888d50d7 | <-- still very slow!!!                                                                    |
| `delete_indices`: reduce looping                                        | ~0.75sec    | ~50sec       | 8a3291b6310d3e4c79eae2f5c709d86e619895b9 | for 20k, loop is reduced from 4000 to 1224                                                |
| `delete_indices`: reduce looping + time comparison | ~0.55sec    | ~5sec        | cc2e3247f13b83a6d2be7f74250b9c92b5541edf |  tiny sacrifice on precision (< 0.01ns on 64 bit OS)                                                                                         |
| remove `delete_indices` logic                                           | ~0.55sec    | ~5sec        | N/A                                      | the lower bound on `delete_indices` optimization, removing it will break corner cases. |

**Comment by pllim** [comment]
Please add change log. I think this is performance bug, so can be under bugfix. Thanks!

**Comment by orionlee** [comment]
A side note on performance improvement option I just noticed that is out of scope of this PR, but I'm mentioning to see if we should open a separate issue to track / tackle it.

If [`bottleneck`](https://github.com/pydata/bottleneck)'s `nanmedian` is used in the test, the running time is further reduced (200k timeseries: from ~5sec to ~1.6sec ; 20k series from ~0.5sec to 0.2sec).

[`astropy.stats.sigma_clip()`](https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clip.html#astropy.stats.sigma_clip) API is set such that by default if `bottleneck` is installed, it will be used (and typically result in reduced running time) to calculate median/mean, similar approach can be used here.

See: https://docs.astropy.org/en/stable/stats/index.html#performance-tips

**Comment by dhomeier** [comment]
> If [`bottleneck`](https://github.com/pydata/bottleneck)'s `nanmedian` is used in the test, the running time is further reduced (200k timeseries: from ~5sec to ~1.6sec ; 20k series from ~0.5sec to 0.2sec).
> 
> [`astropy.stats.sigma_clip()`](https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clip.html#astropy.stats.sigma_clip) API is set such that by default if `bottleneck` is installed, it will be used (and typically result in reduced running time) to calculate median/mean, similar approach can be used here.

Very much in favour of this, if it can be set up the same way for an optional dependency!

**Comment by pllim** [comment]
Why does bottleneck brings back bad vague memories... Maybe @larrybradley remembers...

**Comment by pllim** [comment]
> Does GH Actions honour `[ci skip]`?

Yes, they do now. They skip more things than I like but something is better than nothing.

**Comment by dhomeier** [comment]
> * For the idea of setting something like `1e-6 * (ts_sorted.time[1:]-ts_sorted.time[:-1]).min()` as the epsilon to further reduce looping, if we proceed, what would be a test to ensure we don't screw up?

I don't think the + epsilon trick is needed any more, as this seems to be already as far optimised as could be expected.
Even without that a specific test would still be nice, but I could not come up with anything effective.
Take a time sequence
```python
t = Time(1.41804e9 + np.arange(10000) * 1e-5, format="unix")
flux = np.zeros(10000)
flux[1::100] = 100
ts = TimeSeries(time=t2, data=dict(flux=flux))
ts_b = aggregate_downsample(ts, time_bin_start=ts.time[0] - 0.2*u.us, time_bin_size=1*u.ms)
print(ts_b['flux'].min(), ts_b['flux'].max())
1.0 1.0
```
will print min/max of 100/101 and 100/99 instead for `time_bin_start` offsets of less than some 0.2 μs, since some bins will have one point more or less due to precision limits. But it is not as specific as I'd like it to be, since the accuracy seems the same even when using `ts.time.jd` for the bin checks – although every 3-4 points in `ts` are within 25 μs and identical in `jd`.
So I'd trust this to be robust enough, but pinging @barentsen for thoughts.

**Review by dhomeier — APPROVED** [review]
Approving, but leaving this open for a second reviewer or @astrofrog/@bsipocz to comment on coding style or precision/resolution/testing issues I may have overlooked.

**Comment by orionlee** [comment]
I've done one more sanity test using real data from TESS.
The performance gain is not as substantial: it is still 2-3 times slower than astropy 4.3.1, but is still about 10X better than the released 5.0.4

With TESS Lightcurve data (TIC 329351474, sector 48 (~19k samples)

|                     | `flux` and `flux_err` only | all columns (23) |
| ------------------- | ----------------------- | ---------------- |
| astropy 4.3.1       | ~0.35sec                | ~3.27sec         |
| astropy 5.0.4       | ~9.9sec                 | ~18.5sec         |
| astropy 5.1dev + PR (cc2e3247f13b83a6d2be7f74250b9c92b5541edf) | ~0.93sec                | 8.6sec           |

**Comment by pllim** [comment]
Re: TESS -- Timing aside, are the answers at least more correct with `astropy` v5? I would say slightly slower run for better result is a good trade-off?

File: astropy/timeseries/downsample.py
Comment (mhvk) [reply astropy/timeseries/downsample.py:52]: I think this is meant to return `longdouble` - 80 bits on many systems.  But really should look more carefully before commenting (sorry no time until next week)

**Comment by larrybradley** [comment]
@pllim The issue you are remembering about `bottleneck` was that it could not be installed with Py 3.10 for a long time because the maintainers did not create wheels.  That's no longer an issue and there are now more `bottleneck` maintainers to help with releases.

**Comment by orionlee** [comment]
- For the possibility of using `bottleneck`, I created a separate issue #13078 for it (and a few other ideas on the API level).

- For TESS data test: I don't know if the slow down is the necessary sacrifice to get better result, or is something we could chip off. Anyway, for the purpose of this PR, I don't think it's a showstopper: The main purpose of the test with TESS data is to validate that the PR would improve running time with real data (and it did).

**Comment by orionlee** [comment]
> ... Even better, the `_searchsorted` function can be dropped into https://github.com/astropy/astropy/pull/12703's function_helpers almost as is to speed up np.searchsorted to barely twice the time of running it directly on the ndarrays, so after that enhancement is merged, that can be put into a follow-up PR to speed up all Time objects!

@dhomeier For the possibly dropping a (generalized version) of the `_searchsorted` function to `function_helpers` for all `Time` objects, there is a small caveat: the implementation places some limitation of the precision of Time, as it is comparing seconds in `np.longdouble` (< 0.01ns on 64bit OS). I don't know if such limitation would break some use cases.

**Comment by dhomeier** [comment]
> @dhomeier there is a small caveat in case it's lost in the discussion: the implementation places some limitation of the precision of Time, as it is comparing seconds in `np.longdouble` (< 0.01ns on 64bit OS). I don't know if such limitation would break some use cases.

The unit does not really matter (could of course convert it to attoseconds as well), as `float64` can handle down to ~ 1e-300 s either way; this really comes only into play when combined with longer time ranges where the extended precision of the `Time` object would be fully used. If you can come up with a case where this shows up, that would actually make a good test (even if this PR did not pass at full precision, that could give at least a handle on the limits).
But in general the point of course holds, can we replace the previous, extremely slow implementation of `np.searchsorted` with a 1000 x faster one that might compromise on accuracy in some cases, and should we still provide the alternative at best precision. That might be done by adding a `searchsorted` to `Time` that would, similarly to its `argsort`, use the full accuracy of the `jd1` and `jd2` components where required. That would still change the default behaviour of `np.searchsorted(Time)`, though for me I'd rather expect that form to have similar performance to other numpy array functions, and at the precision of the standard array dtypes. Meaning I find it more intuitive, even if it constitutes a change from the current behaviour.

**Comment by orionlee** [comment]
I've done all I planned to do, and made one more commit.

|                                                       | 20k samples | 200k samples | Commit                                   | Comment                                                                                   |
| ----------------------------------------------------- | ----------- | ------------ | ---------------------------------------- | ----------------------------------------------------------------------------------------- |
| astropy 4.3.1                                         | ~0.42sec    | ~4.2sec      | --                                      |                                                                                           |
| astropy 5.0.4                                         | ~10sec      | ~252sec      | --                                      |                                                                                           |
| `searchsorted()` fix (this PR initially)            | ~1.3sec     | 147sec       | a753502f60fab3e3631c8b0e69efc4a0888d50d7 | <-- still very slow!!!                                                                    |
| `delete_indices`: reduce looping                   | ~0.75sec    | ~50sec       | 8a3291b6310d3e4c79eae2f5c709d86e619895b9 | for 20k, loop is reduced from 4000 to 1224                                                |
| `delete_indices`: reduce looping + time comparison | ~0.55sec    | ~5.2sec      | cc2e3247f13b83a6d2be7f74250b9c92b5541edf |                                                                                           |
| use relative time when possible                       | ~0.52sec    | ~4.9sec      | a50ea29352befc8c3f1387d6ebf9d3d3565be516 | <-- lastest PR result                                                                     |
| remove `delete_indices` logic                      | ~0.55sec    | ~5sec        | N/A                                      | the lower bound on `delete_indices` optimization, removing it will break corner cases. |


---
- In the latest commit, I made the remaining `Time` operations use relative time as well (and pass relative time to `np.searchsorted()` directly). The speed gain is small. But I think it's easier to understand/maintain, as relative times are now consistently used throughout the codes.

- The `epsilon` idea is abandoned for 2 reasons:
  i. the potential gain is negligible, number times to loop (`len(noncontiguous_bins_indices)`) < 30 in cases I've test (20k artificial timeseries, 200k artificial timeseries, 19k TESS data)
  ii. adding the `epsilon` logic also broke an unit test. Since the potential gain is negligible I did not chase down any further.

- The profile result of the PR shows the biggest chunk of time spend outside of `reduceat` / `np.nanmean` is `Column.__array_finalize__`.
  - I don't know how much of it is inherent overhead of using `Table`, or if there are code paths in `arrgegate_downsample()` that inflated the overhead.


Profile result highlights (binning a 200k timeseries):

| ncalls | cumtime | % cumtime | function                                         |
| ------ | ------- | --------- | ------------------------------------------------ |
| 1      | 7.758   | 100.0%    | `downsample.py:48(aggregate_downsample)`          |
| 1      | 6.819   | 87.9%     | `downsample.py:16(reduceat)`                       |
| 40000  | 5.558   | 71.6%     | `<__array_function__ internals>:177(nanmean)` |
| 160048 | 3.968   | 51.1%     | `column.py:569(__array_finalize__)`           |

**Comment by orionlee** [comment]
@dhomeier The following is probably not quite the test on the limitation you have in mind, but it does show the limitation of the relative time conversion in the PR.

On 64 bit Windows, the limit is around `0.008ns`. 
```python
from astropy.time import Time
from astropy import units as u
import numpy as np

def _to_relative_longdouble(time: Time, rel_base: Time) -> np.longdouble:
    return (time - rel_base).to_value(format="sec", subfmt="long")

t_base = Time(2457000, format="jd")
t2 = t_base + 1 *u.s
t3 = t2 + 0.008 *u.ns  # 0.009 ns would work

r_t2 = _to_relative_longdouble(t2, t_base)
r_t3 = _to_relative_longdouble(t3, t_base)

# r_t3 > r_t2 == False when their difference reach the limit of the precision.
print(r_t3 > r_t2, r_t3, r_t2)
```
Output: 
```python
False 0.9999999999969589 0.9999999999969589
```

**Comment by dhomeier** [comment]
I can confirm this (`t3 > t2`) on 64 bit macOS as well; the interesting part is that this is not a limit of floating point precision, but of `Time` arithmetic after all, because I also get
```python
t3 = t2 + 0.008 *u.ns
t3 - t2
<TimeDelta object: scale='tai' format='jd' value=0.0>
```
which is at first glance surprising, since the subtraction is indeed using the `jd1` and `jd2` components for maximum precision, but the scale of the output actually gives a hint: in `Time.__sub__`
https://github.com/astropy/astropy/blob/b54cc0d4e9b749384126305c63f1597b69cbb37c/astropy/time/core.py#L2166
both inputs are first converted to a common, different scale (`tai`), and on this scale the `jd2` happens to have larger values, with the difference below the relative accuracy:
```python
print(t2._time.jd2, t3._time.jd2)
1.1574071128589436e-11 1.157411232827199e-11
print(t2.tai._time.jd2, t3.tai._time.jd2)
0.00040509260416665693 0.00040509260416665693
```
So I think you could actually improve the accuracy of your `_to_relative_longdouble` in most cases a bit further by checking first
```python
    if time.scale == rel_base.scale:
        return ((time._time.jd1 - rel_base._time.jd1) + (time._time.jd2 - rel_base._time.jd2)) * 86400
```
Wondering if `Time.__sub__` should do a similar thing and at least check if both inputs are already on identical scales before converting them both. But then, this is maybe pushing things, as I also find
```python
t3 = t2 + 0.008 *u.ns
t4 = t2 + 0.004 *u.ns
t4 > t3
True
```
– inspecting the `jd2` values for any such offset in the picosecond range shows that they take quite random numbers and are definitely at the precision limit even within their scale, so there is probably no point in putting too much further effort into this.

**Comment by mhvk** [comment]
This is all looking quite exciting! I think it makes sense to try to squeeze the last bit of accuracy and performance in a separate PR. I would think, though, that the real test case would be one where one has an array of `Time` that consists of clusters of times very close to each other, but with the clusters far away (say, decades apart while the times are within a few ns).

**Comment by dhomeier** [comment]
Yes; ns resolution on decades timescales will definitely exceed `float64` precision (to my knowledge numpy does not provide direct use of 80 bit data types). Might be possible to implement some kind of two-stage search using first `jd1` and then `jd2`, but such cases sound exotic enough to perhaps best defer simply to the slow `Time` implementation.

**Comment by orionlee** [comment]
> I would think, though, that the real test case would be one where one has an array of Time that consists of clusters of times very close to each other, but with the clusters far away (say, decades apart while the times are within a few ns).

@mvhk This is a good point -  it's probable to combine current observations with archival data that could be a few years old.

The following is the limitation on the current implementation on 64 bit Windows (limited by `Time` subtraction rather than `float64` as @dhomeier pointed out).

| Cluster difference | precision limit |
| ------------------ | --------------- |
| 1 year             | ~1.8ns          |
| 10 years           | ~29ns           |
| 50 years           | ~119ns          |
| 100 years          | ~238ns          |


For the purpose of `aggregate_downsample()`, are such limits acceptable?


Script to show the limit on relative time.
```python
# the cluster difference
diff_from_base = 100 * u.year
diff_t2_t3 = 238 * u.ns

from astropy.time import Time
from astropy import units as u
import numpy as np

def _to_relative_longdouble(time: Time, rel_base: Time) -> np.longdouble:
    return (time - rel_base).to_value(format="sec", subfmt="long")

t_base = t_base = Time('1970-01-01T12:30:31.000')
t2 = t_base + diff_from_base
t3 = t2 + diff_t2_t3

r_t2 = _to_relative_longdouble(t2, t_base)
r_t3 = _to_relative_longdouble(t3, t_base)

# r_t3 > r_t2 == False when their difference reaches the limit of the precision.
print(r_t3 > r_t2, r_t3, r_t2)
```

**Comment by orionlee** [comment]
I identified a separate performance issue (i.e., not a regression) with a fix. Let me know if we should bundle it to this PR or handle it separately.
- The current `aggregate_downsample()` is relatively slow when it is binning a column that is not a `Quantity`, e.g., a plain `Column`.
  - Essentially, when `reduceat()` operates on a `Column` rather than `ndarray` (the code path used for `Quantity`), it incurs significant overhead.
- it does affect the numbers I reported in this PR: if we fix the issue, the ~4.9sec time  reported for 200k samples  can be further reduced to ~2sec.

---

One fix would be to introduce a code path to handle `Column` separately. In
https://github.com/astropy/astropy/blob/a50ea29352befc8c3f1387d6ebf9d3d3565be516/astropy/timeseries/downsample.py#L237-L241

Add an extra case so that `reduceat` operates on plain `ndarray`:
```python
        elif isinstance(values, Column):
            data = np.ma.zeros(n_bins, dtype=values.dtype)
            data.mask = 1
            data[unique_indices] = type(values)(reduceat(values.value, groups, aggregate_func),
                                          dtype=values.unit, copy=False)
            data.mask[unique_indices] = 0
```

I have not considered other MixIn column types yet.

**Comment by orionlee** [comment]
@mvhk @dhomeier in case the [previous message](https://github.com/astropy/astropy/pull/13069#issuecomment-1093069184) gets lost. I'd like to double check to see if the time precision limit of this PR is acceptable. 

The following is the precision limitation on the current implementation on 64 bit Windows (limited by `Time` subtraction rather than `float64` as @dhomeier pointed out).

| Cluster difference | precision limit |
| ------------------ | --------------- |
| 1 year             | ~1.8ns          |
| 10 years           | ~29ns           |
| 50 years           | ~119ns          |
| 100 years          | ~238ns          |

Outside of feedback from other reviewers, I think the precision limit might possibly be a showstopper for this PR to be merged.

**Comment by dhomeier** [comment]
I am rather leaning to accept with that precision **unless** another reviewer yet objects –

1. < μs on a multi-decade timescale sounds good enough to me, as I think photographic exposures are rarely recorded at higher accuracy ;-)
2. My impression from the tests above is that those numbers are very close to the limits of time arithmetic either way, so getting below, say 29 ns at a very considerable performance penalty seems hardly worth it when time deltas may no longer be accurate at 10 ns anyway.

**Comment by orionlee** [comment]
> > I am rather leaning to accept with that precision **unless** another reviewer yet objects – ...
> 
> When an agreement is reached, I'd like to add a test case that explicitly tests the agreed upon precision limit, as a form of "enforceable" documentation.

A test case has been added, with a strawman precision limit of 500ns.

**Comment by orionlee** [comment]
@pllim what could be done to get this PR moving?

**Comment by pllim** [comment]
Thanks for the ping and sorry this fell through the cracks. Looks like Derek approved it a while ago but would be nice to have the timeseries maintainers have a look too: @astrofrog @bsipocz

**Comment by pllim** [comment]
In the meantime, rebasing would bring your branch up-to-date with `main` and reset the stale clock. FYI.

**Comment by dhomeier** [comment]
> Looks like Derek approved it a while ago but would be nice to have the timeseries maintainers have a look too: @astrofrog @bsipocz

Thanks also for the added test! I've been wondering if one might even better scale the precision relative to `diff_to_base` (I succeeded with `precision_limit` [5, 50, 250, 500] * u.ns, would have to be confirmed on other platforms), and also if this could be tested more directly on the bins returned from an `aggregate_downsample` call. But holding direct suggestions until you've rebased.

tldr for the maintainers: functionally this looks all good; the unresolved comments are about implementation and code style:
1. does `_to_relative_longdouble` warrant an extra function?
2. are single-use variables like `noncontiguous_bins_indices` beneficial for readability?

File: astropy/timeseries/tests/test_downsample.py
Comment (dhomeier) [inline astropy/timeseries/tests/test_downsample.py:201]:
```suggestion
@pytest.mark.parametrize("scale",
                         [1, 10, 50, 100])
def test_time_precision_limit(scale):
```
Might set scaled limits (see below).

File: astropy/timeseries/tests/test_downsample.py
Comment (dhomeier) [inline astropy/timeseries/tests/test_downsample.py:221]:
```suggestion
    precision_limit = scale * 5 * u.ns

    from astropy.timeseries.downsample import _to_relative_longdouble

    t_base = Time('1980-01-01T12:30:31.000', format='isot', scale='tdb')
    t2 = t_base + scale * u.year
    t3 = t2 + precision_limit
```

**Comment by orionlee** [comment]
>  I've been wondering if one might even better scale the precision relative to diff_to_base (I succeeded with precision_limit [5, 50, 250, 500] * u.ns, consistent with what you found before, but would have to be confirmed on other platforms) , and also if this could be tested more directly on the bins returned from an aggregate_downsample call. But holding further suggestions until you've rebased.

@dhomeier For the suggested changes to the test to make the `precision limit` scaled relative to `diff_to_base`, I suggest we defer it until a decision is made with the timeseries maintainers (the 500ns limit used in the test is a strawman).
1. I think the test should document what the `precision_limit` `aggregate_downsample()` guarantees, irrespective of how the implementation changes.
2. A `precision_limit` scaled relative to `diff_to_base` smells too close to implementation to me.
3. I do realize, however, the test is testing implementation details , rather than `aggregate_downsample()`'s behavior (the returned bins as you said). I haven't yet figured out how to do it without making the test cryptic. Ideas are welcome!

**Comment by pllim** [comment]
What is the status of this? Looks like there is some unresolved discussions?

**Review by astrofrog — APPROVED** [review]
Apologies for not getting to this sooner - in my view, <µs precision is sufficient specifically in the context of ``aggregate_downsample``. I can't think of an actual science case where this would not be enough - even for say pulsars which have periods down to the ~ms level, such a precision would be much much smaller than the period.

However I think we should just be upfront and mention the precision limits in the docstring for the function? I am marking as approved to not hold this up in case I don't have time to re-review but I think the only thing to do before merging would be to add the note in the docstring.

File: astropy/timeseries/downsample.py
Comment (dhomeier) [reply astropy/timeseries/downsample.py:45]: On x86_64 with numpy 1.23 this does return np.longdouble (np.float128).
Might still save the trouble by just using `(time - rel_base).to_value(format="sec", subfmt="long")` inline, I think.

**Review by dhomeier — APPROVED** [review]
I agree stating the expected precision clearly in the docstring is a good solution to this – as documented in https://github.com/astropy/astropy/pull/13069#issuecomment-1093069184.
I think this should also make clear that this has the characteristics of a relative precision.

**Comment by dhomeier** [comment]
@orionlee do you have any comments on the open points in the discussion, or would you rather prefer this to be merged as is (which I think would be good enough)? Not sure if this can still pass as a bugfix, but to be on the safe side I think it would be better to meet the deadline for merging 5.2 contributions this Friday.

**Comment by dhomeier** [comment]
@astrofrog @bsipocz @pllim seems there has not been any GH activity from @orionlee for several months; can we merge as is and deal with remaining issues separately?

**Comment by pllim** [comment]
Would at least need a rebase and make sure the CI failures are unrelated.

As for merging as is or not, that is up to timeseries maintainers. I am not qualified to make that call. Thanks!

**Comment by astrofrog** [comment]
I have rebased this, we can merge once/if CI passes

**Comment by orionlee** [comment]
Thanks all for resolving this issue!

@pllim This PR is merged after v5.1.1 . Does that mean it needs to be backported to v5.1 as well?

I check the version history of v5.1.x branch, the PR is yet included:

https://github.com/astropy/astropy/blob/v5.1.x/astropy/timeseries/downsample.py

**Comment by pllim** [comment]
There won't be another 5.1.x series, so this will be in 5.2 and 5.0.6.

**Pull request merged**
